### ISSUE TITLE
parallel query call not supported

### ISSUE THREAD

**Issue by mortifia** [porsager/postgres#142]
this.db = postgres(url, {conf});

``` js

    async coucou() { // for test actually 
        console.log("coucou");

        const res = await this.db`
        select *
        from "user".users
        `
        console.log("coucou end");
        return res[0];
    }
    async auth(token_uuid) { // for test actually 
        console.log("auth");
        const res = await this.db`
        select token_uuid, user_uuid, created_on, renewable, blocked
        from authentication.jwt_token
        where token_uuid = ${token_uuid}
        `
        console.log("auth end");
    }
```

``` 
🚀  Server ready at http://localhost:4000/
auth
coucou
auth end
coucou end
auth
coucou
auth end
coucou end
auth
coucou
auth end
coucou end
auth
coucou
(node:17536) UnhandledPromiseRejectionWarning: Error: failed password authentication for user   xxxxxxx
    at Object.generic (\node_modules\postgres\lib\types.js:200:5)
    at Object.ErrorResponse (\node_modules\postgres\lib\backend.js:125:22)
    at Socket.data (\node_modules\postgres\lib\connection.js:167:25)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:309:12)
    at readableAddChunk (_stream_readable.js:284:9)
    at Socket.Readable.push (_stream_readable.js:223:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:17536) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:17536) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, 
promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.  
coucou end
auth
coucou
auth end
coucou end
auth
coucou
auth end
coucou end
auth
coucou
coucou end
auth end
auth
coucou
coucou end
auth end
auth
coucou
coucou end
\node_modules\postgres\lib\backend.js:175
    if (backend.query.result.command) {
                      ^

TypeError: Cannot read property 'result' of undefined
    at Object.RowDescription (\node_modules\postgres\lib\backend.js:175:23)
    at Socket.data (\node_modules\postgres\lib\connection.js:167:25)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:309:12)
    at readableAddChunk (_stream_readable.js:284:9)
    at Socket.Readable.push (_stream_readable.js:223:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! test-apollo-server@1.0.0 start: `node index.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the test-apollo-server@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.     

npm ERR! A complete log of this run can be found in:
npm ERR!     \AppData\Roaming\npm-cache\_logs\2021-01-12T22_17_23_588Z-debug.log
```

```
🚀  Server ready at http://localhost:4000/
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth end
auth end
auth end
auth
coucou
auth
coucou
auth
coucou
auth
coucou
auth end
auth
coucou
\node_modules\postgres\lib\backend.js:175
    if (backend.query.result.command) {
                      ^

TypeError: Cannot read property 'result' of undefined
    at Object.RowDescription (\node_modules\postgres\lib\backend.js:175:23)
    at Socket.data (\node_modules\postgres\lib\connection.js:167:25)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:309:12)
    at readableAddChunk (_stream_readable.js:284:9)
    at Socket.Readable.push (_stream_readable.js:223:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! test-apollo-server@1.0.0 start: `node index.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the test-apollo-server@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     \AppData\Roaming\npm-cache\_logs\2021-01-12T23_44_06_752Z-debug.log
```

**Comment by mortifia** [comment]
db and the test in the same server

**Comment by marcbachmann** [comment]
Are you also using `SCRAM-SHA-256` authentication? #123

**Comment by marcbachmann** [comment]
Wow, took me 10 minutes to find the issue 😃 
Pull request coming.

**Comment by mortifia** [comment]
url: 'postgres://xxxx:xx@127.0.0.1:5432/xx',
connection: {
        host: '127.0.0.1',
        port: '5432',
        user: 'xxxx',
        password: 'xx',
        database: 'xx',
      }
      
      
      
postgresql 13
```
# TYPE  DATABASE        USER            ADDRESS                 METHOD

# "local" is for Unix domain socket connections only
local   all             all                                     scram-sha-256
# IPv4 local connections:
host    all             all             127.0.0.1/32            scram-sha-256
# IPv6 local connections:
host    all             all             ::1/128                 scram-sha-256
# Allow replication connections from localhost, by a user with the
# replication privilege.
local   replication     all                                     scram-sha-256
host    replication     all             127.0.0.1/32            scram-sha-256
host    replication     all             ::1/128                 scram-sha-256
```

**Comment by marcbachmann** [comment]
Fixed in https://github.com/porsager/postgres/pull/145. Could you test it?

**Comment by mortifia** [comment]
stress test basic 1800 request/s (user->api)


load stress test lib (fail) 

``` js
async coucou() {
        console.log("coucou");

        /*const res = await this.db`
        select *
        from "user".users
        `*/

        const tmp = await Promise.allSettled([
            this.db`select * from "user".users`,
            this.db`select * from "user".users`])
        console.log(tmp)
        console.log("coucou end");

        return res[0].value[0];
    }
```


log

``` 

> test-apollo-server@1.0.0 start 
> node index.js

­ƒÜÇ  Server ready at http://localhost:4000/
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'rejected',
    reason: Error: authentification par mot de passe ´┐¢chou´┐¢e pour l'utilisateur  ´┐¢ xxxx ´┐¢
        at Object.generic (node_modules\postgres\lib\types.js:200:5)
        at Object.ErrorResponse (node_modules\postgres\lib\backend.js:125:22)
        at Socket.data (node_modules\postgres\lib\connection.js:167:25)
        at Socket.emit (events.js:315:20)
        at addChunk (_stream_readable.js:309:12)
        at readableAddChunk (_stream_readable.js:284:9)
        at Socket.Readable.push (_stream_readable.js:223:10)
        at TCP.onStreamRead (internal/stream_base_commons.js:188:23) {
      severity_local: 'FATAL',
      severity: 'FATAL',
      code: '28P01',
      file: 'd:\\pginstaller_13.auto\\postgres.windows-x64\\src\\backend\\libpq\\auth.c',
      line: '334',
      routine: 'auth_failed'
    }
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou
[
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  },
  {
    status: 'fulfilled',
    value: [ [Object], [Object], [Object], count: 3, command: 'SELECT' ]
  }
]
coucou end
coucou


node_modules\postgres\lib\backend.js:175
    if (backend.query.result.command) {
                      ^

TypeError: Cannot read property 'result' of undefined
    at Object.RowDescription (node_modules\postgres\lib\backend.js:175:23)
    at Socket.data (node_modules\postgres\lib\connection.js:167:25)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:309:12)
    at readableAddChunk (_stream_readable.js:284:9)
    at Socket.Readable.push (_stream_readable.js:223:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! test-apollo-server@1.0.0 start: `node index.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the test-apollo-server@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     npm-cache\_logs\2021-01-19T00_07_58_632Z-debug.log
```

**Comment by mortifia** [comment]
package-lock.json

    "postgres": {
      "version": "1.0.2",
      "resolved": "https://registry.npmjs.org/postgres/-/postgres-1.0.2.tgz",
      "integrity": "sha512-zeLgt42KSUNgX/uvo+gbVxTAYwgSY6MIKuU/a8YWuObX4rtGuKrVWopvEAqIAPSO0FeHS1TsSKnqPjoufPy8NA=="
    },

**Comment by marcbachmann** [comment]
the fix isn't released yet. Did you test against #145?

**Comment by marcbachmann** [comment]
That one doesn't report any errors anymore with the new code:
```js
async function start (port) {
  const sql = require('./')(`postgres://postgres:password@localhost:${port}/postgres`, {max: 100})

  try {
    const promises = []
    for (let index = 0; index < 10000; index++) {
      promises.push(sql`SELECT 1`)
    }
    await Promise.all(promises)
  } finally {
    sql.end({timeout: 0})
  }
}

start(5433)
  .catch((err) => console.error('Error from instance with scram authentication: ', err))

```

**Comment by mortifia** [comment]
with knex .... (not good)
- user -> api 900/s
- api -> database 900/s
- after the first test I immediately forgot it consumes too much

with pg-pool (npm i pg)
- user -> api 1800/s
- api -> database 5200/s ( with  Promise.allSettled )
- 12ms
- I currently use it

with postgres
- user -> api 1800/s 2000/s
- api -> database (test 1 with comments) 1800/s 2000/s
api -> database (test 2 with  Promise.allSettled ) on test  (not crash all good)

**Comment by mortifia** [comment]
now that it is repaired I will see to see that apollo-server adds the lib created with it in the Community data sources documentation

when the patch will be on npm

**Comment by porsager** [comment]
Awesome work, and good find @marcbachmann ! That's a pretty obvious issue in hindsight 😃

**Comment by porsager** [comment]
Released as [v2.0.0-beta.3](https://github.com/porsager/postgres/releases/tag/v2.0.0-beta.3)

**Comment by porsager** [comment]
Please reopen if this is still an issue ;)
